File: c++/day25/part1.cpp
#include "part1.h"

using namespace std;


namespace Day25 {

    vector<string> parse(const string& filename) {
        return getFileLines(filename);
    }


    namespace Part1 {

        long long snafuToDecimal(const string& snafu) {
            long long result = 0;
            for (int i = 0; i < snafu.size(); ++i) {
                int digit = 0;
                if (snafu[i] == '2')      { digit  = 2; }
                else if (snafu[i] == '1') { digit  = 1; }
                else if (snafu[i] == '-') { digit = -1; }
                else if (snafu[i] == '=') { digit = -2; }
                result = 5 * result + digit;
            }
            return result;
        }

        string decimalToSnafu(const long long n) {
            long long curr = n;
            string result {};
            const map<int, string> snafuDigits { {-2, "="}, {-1, "-"}, {0, "0"}, {1, "1"}, {2, "2"} };
            while (curr > 0) {
                int remainder = curr % 5;
                if (remainder > 2) {
                    remainder -= 5;
                }
                curr = (curr - remainder) / 5;
                result = snafuDigits.at(remainder) + result;
            }
            return result;
        }

        string solve(const vector<string>& snafus) {
            long long result = 0;
            for (const auto& snafu : snafus) {
                result += snafuToDecimal(snafu);
            }
            return decimalToSnafu(result);
        }

    }
}

File: python/day25/part1.py
#!/usr/bin/env python3
import time

DEC_TO_STAFU = {'0': 0, '1': 1, '2': 2, '-': -1, '=': -2}
STAFU_TO_DEC = {0: '0', 1: '1', 2: '2', -1: '-', -2: '='}


def to_decimal(snafu):
    return sum([DEC_TO_STAFU[c] * pow(5, len(snafu)-1-i) for (i, c) in enumerate(snafu)])


def to_snafu(number):
    res = ""
    while number > 0:
        remainder = number % 5
        if remainder > 2:
            remainder -= 5
        res = STAFU_TO_DEC[remainder] + res
        number -= remainder
        number //= 5
    return res


def parse(input_path):
    with open(input_path, 'r') as f:
        return [line.strip() for line in f.readlines()]


def solve_part1(snafus):
    return to_snafu(sum([to_decimal(snafu) for snafu in snafus]))


if __name__ == "__main__":
    start = time.time()
    parsed = parse("data.txt")
    result = solve_part1(parsed)
    print(f'Part 1 : {result}  ({round(1000 * (time.time() - start))} ms)')
